File: src-tauri/src/omarchy.rs
//! Omarchy theme integration: read the active theme's colors and watch for
//! theme switches, emitting `omarchy:theme-changed` to the frontend.
//!
//! The active theme lives at `~/.config/omarchy/current/theme`, a symlink that
//! `omarchy-theme-set` swaps atomically. We watch the parent directory
//! `~/.config/omarchy/current` (non-recursively), debounce, re-read, and emit.

use std::path::PathBuf;
use std::sync::mpsc;
use std::time::Duration;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter};

/// Theme colors mirrored to the frontend as CSS custom properties.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OmarchyTheme {
    pub name: String,
    pub accent: String,
    pub foreground: String,
    pub background: String,
    pub cursor: String,
    pub selection_foreground: String,
    pub selection_background: String,
    /// color0..color15 as `#rrggbb`.
    pub palette: Vec<String>,
}

/// Raw `colors.toml` shape as written by omarchy themes.
#[derive(Debug, Deserialize)]
struct ColorsToml {
    accent: Option<String>,
    foreground: Option<String>,
    background: Option<String>,
    cursor: Option<String>,
    selection_foreground: Option<String>,
    selection_background: Option<String>,
    color0: Option<String>,
    color1: Option<String>,
    color2: Option<String>,
    color3: Option<String>,
    color4: Option<String>,
    color5: Option<String>,
    color6: Option<String>,
    color7: Option<String>,
    color8: Option<String>,
    color9: Option<String>,
    color10: Option<String>,
    color11: Option<String>,
    color12: Option<String>,
    color13: Option<String>,
    color14: Option<String>,
    color15: Option<String>,
}

/// `~/.config/omarchy/current`.
fn current_dir() -> Option<PathBuf> {
    dirs::config_dir().map(|d| d.join("omarchy").join("current"))
}

/// Built-in kanagawa fallback used when the omarchy theme files are missing.
pub fn kanagawa_fallback() -> OmarchyTheme {
    OmarchyTheme {
        name: "kanagawa".to_string(),
        accent: "#7e9cd8".to_string(),
        foreground: "#dcd7ba".to_string(),
        background: "#1f1f28".to_string(),
        cursor: "#c8c093".to_string(),
        selection_foreground: "#c8c093".to_string(),
        selection_background: "#2d4f67".to_string(),
        palette: vec![
            "#090618", "#c34043", "#76946a", "#c0a36e", "#7e9cd8", "#957fb8", "#6a9589", "#c8c093",
            "#727169", "#e82424", "#98bb6c", "#e6c384", "#7fb4ca", "#938aa9", "#7aa89f", "#dcd7ba",
        ]
        .into_iter()
        .map(String::from)
        .collect(),
    }
}

/// Read the active omarchy theme, falling back to kanagawa on any problem.
pub fn read_theme() -> OmarchyTheme {
    match try_read_theme() {
        Ok(theme) => theme,
        Err(err) => {
            tracing::debug!(error = %err, "falling back to kanagawa theme");
            kanagawa_fallback()
        }
    }
}

fn try_read_theme() -> Result<OmarchyTheme> {
    let dir = current_dir().context("no config dir")?;
    let colors_path = dir.join("theme").join("colors.toml");
    let name_path = dir.join("theme.name");

    let raw = std::fs::read_to_string(&colors_path)
        .with_context(|| format!("reading {}", colors_path.display()))?;
    let colors: ColorsToml =
        toml::from_str(&raw).with_context(|| format!("parsing {}", colors_path.display()))?;

    let name = std::fs::read_to_string(&name_path)
        .map(|s| s.trim().to_string())
        .unwrap_or_else(|_| "unknown".to_string());

    let fb = kanagawa_fallback();
    let palette: Vec<String> = [
        colors.color0,
        colors.color1,
        colors.color2,
        colors.color3,
        colors.color4,
        colors.color5,
        colors.color6,
        colors.color7,
        colors.color8,
        colors.color9,
        colors.color10,
        colors.color11,
        colors.color12,
        colors.color13,
        colors.color14,
        colors.color15,
    ]
    .into_iter()
    .enumerate()
    .map(|(i, c)| c.unwrap_or_else(|| fb.palette[i].clone()))
    .collect();

    Ok(OmarchyTheme {
        name,
        accent: colors.accent.unwrap_or(fb.accent),
        foreground: colors.foreground.unwrap_or(fb.foreground),
        background: colors.background.unwrap_or(fb.background),
        cursor: colors.cursor.unwrap_or(fb.cursor),
        selection_foreground: colors
            .selection_foreground
            .unwrap_or(fb.selection_foreground),
        selection_background: colors
            .selection_background
            .unwrap_or(fb.selection_background),
        palette,
    })
}

/// Spawn a background thread that watches for omarchy theme changes and emits
/// `omarchy:theme-changed`. Runs for the lifetime of the app.
pub fn spawn_watcher(app: AppHandle) {
    std::thread::spawn(move || {
        if let Err(err) = watch_loop(app) {
            tracing::warn!(error = %err, "omarchy watcher stopped");
        }
    });
}

fn watch_loop(app: AppHandle) -> Result<()> {
    use notify::{RecursiveMode, Watcher};

    let dir = match current_dir() {
        Some(d) if d.exists() => d,
        _ => {
            tracing::info!("omarchy current dir not present; theme watcher inactive");
            return Ok(());
        }
    };

    let (tx, rx) = mpsc::channel();
    let mut watcher = notify::recommended_watcher(move |res: notify::Result<notify::Event>| {
        // Ignore access events: our own `read_theme()` reads files inside the
        // watched dir, so reacting to reads creates a 300ms feedback loop
        // (and any other process reading the theme would retrigger us too).
        match &res {
            Ok(event) if event.kind.is_access() => {}
            _ => {
                let _ = tx.send(res);
            }
        }
    })
    .context("creating fs watcher")?;
    watcher
        .watch(&dir, RecursiveMode::NonRecursive)
        .with_context(|| format!("watching {}", dir.display()))?;

    tracing::info!(dir = %dir.display(), "watching omarchy theme dir");

    // Block for the next event; the channel closes when the watcher is dropped.
    let mut last = read_theme();
    while let Ok(first) = rx.recv() {
        if first.is_err() {
            continue;
        }
        // Debounce: drain any events that arrive within ~300ms.
        while rx.recv_timeout(Duration::from_millis(300)).is_ok() {}

        // Only emit on an actual change; the watcher fires for unrelated
        // churn in the dir (e.g. background rotation touching symlinks).
        let theme = read_theme();
        if theme == last {
            continue;
        }
        last = theme.clone();
        if let Err(err) = app.emit("omarchy:theme-changed", &theme) {
            tracing::warn!(error = %err, "failed to emit theme change");
        } else {
            tracing::info!(theme = %theme.name, "emitted omarchy theme change");
        }
    }
    Ok(())
}

File: internal/omarchy/omarchy.go
// Package omarchy reads the active omarchy theme's colors and watches for
// theme switches.
//
// The active theme lives at ~/.config/omarchy/current/theme, a symlink that
// omarchy-theme-set swaps atomically. We watch the parent directory
// ~/.config/omarchy/current (non-recursively), debounce, re-read, and emit.
package omarchy

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/fsnotify/fsnotify"
	toml "github.com/pelletier/go-toml/v2"

	"cosmicmail/internal/models"
	"cosmicmail/internal/xdg"
)

// Debounce is how long the watcher waits for a burst of events to settle
// before re-reading the theme.
const Debounce = 300 * time.Millisecond

type colorsToml struct {
	Accent              *string `toml:"accent"`
	Foreground          *string `toml:"foreground"`
	Background          *string `toml:"background"`
	Cursor              *string `toml:"cursor"`
	SelectionForeground *string `toml:"selection_foreground"`
	SelectionBackground *string `toml:"selection_background"`
	Color0              *string `toml:"color0"`
	Color1              *string `toml:"color1"`
	Color2              *string `toml:"color2"`
	Color3              *string `toml:"color3"`
	Color4              *string `toml:"color4"`
	Color5              *string `toml:"color5"`
	Color6              *string `toml:"color6"`
	Color7              *string `toml:"color7"`
	Color8              *string `toml:"color8"`
	Color9              *string `toml:"color9"`
	Color10             *string `toml:"color10"`
	Color11             *string `toml:"color11"`
	Color12             *string `toml:"color12"`
	Color13             *string `toml:"color13"`
	Color14             *string `toml:"color14"`
	Color15             *string `toml:"color15"`
}

// CurrentDir is ~/.config/omarchy/current.
func CurrentDir() (string, error) {
	cfg, err := xdg.ConfigDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(cfg, "omarchy", "current"), nil
}

// KanagawaFallback is the built-in theme used when the omarchy files are
// missing.
func KanagawaFallback() models.OmarchyTheme {
	return models.OmarchyTheme{
		Name: "kanagawa", Accent: "#7e9cd8", Foreground: "#dcd7ba", Background: "#1f1f28", Cursor: "#c8c093",
		SelectionForeground: "#c8c093", SelectionBackground: "#2d4f67",
		Palette: []string{
			"#090618", "#c34043", "#76946a", "#c0a36e", "#7e9cd8", "#957fb8", "#6a9589", "#c8c093",
			"#727169", "#e82424", "#98bb6c", "#e6c384", "#7fb4ca", "#938aa9", "#7aa89f", "#dcd7ba",
		},
	}
}

// ReadTheme reads the active omarchy theme, falling back to kanagawa on any
// problem.
func ReadTheme() models.OmarchyTheme {
	dir, err := CurrentDir()
	if err != nil {
		return KanagawaFallback()
	}
	theme, err := ReadThemeFrom(dir)
	if err != nil {
		slog.Debug("falling back to kanagawa theme", "error", err)
		return KanagawaFallback()
	}
	return theme
}

// ReadThemeFrom reads `<dir>/theme/colors.toml` and `<dir>/theme.name`.
func ReadThemeFrom(dir string) (models.OmarchyTheme, error) {
	colorsPath := filepath.Join(dir, "theme", "colors.toml")
	raw, err := os.ReadFile(colorsPath)
	if err != nil {
		return models.OmarchyTheme{}, fmt.Errorf("reading %s: %w", colorsPath, err)
	}
	var colors colorsToml
	if err := toml.Unmarshal(raw, &colors); err != nil {
		return models.OmarchyTheme{}, fmt.Errorf("parsing %s: %w", colorsPath, err)
	}
	name := "unknown"
	if b, err := os.ReadFile(filepath.Join(dir, "theme.name")); err == nil {
		name = strings.TrimSpace(string(b))
	}
	fb := KanagawaFallback()
	pick := func(v *string, fallback string) string {
		if v != nil {
			return *v
		}
		return fallback
	}
	raws := []*string{colors.Color0, colors.Color1, colors.Color2, colors.Color3, colors.Color4, colors.Color5, colors.Color6, colors.Color7,
		colors.Color8, colors.Color9, colors.Color10, colors.Color11, colors.Color12, colors.Color13, colors.Color14, colors.Color15}
	palette := make([]string, 16)
	for i, c := range raws {
		palette[i] = pick(c, fb.Palette[i])
	}
	return models.OmarchyTheme{
		Name:                name,
		Accent:              pick(colors.Accent, fb.Accent),
		Foreground:          pick(colors.Foreground, fb.Foreground),
		Background:          pick(colors.Background, fb.Background),
		Cursor:              pick(colors.Cursor, fb.Cursor),
		SelectionForeground: pick(colors.SelectionForeground, fb.SelectionForeground),
		SelectionBackground: pick(colors.SelectionBackground, fb.SelectionBackground),
		Palette:             palette,
	}, nil
}

// Equal compares two themes.
func Equal(a, b models.OmarchyTheme) bool {
	if a.Name != b.Name || a.Accent != b.Accent || a.Foreground != b.Foreground || a.Background != b.Background || a.Cursor != b.Cursor ||
		a.SelectionForeground != b.SelectionForeground || a.SelectionBackground != b.SelectionBackground || len(a.Palette) != len(b.Palette) {
		return false
	}
	for i := range a.Palette {
		if a.Palette[i] != b.Palette[i] {
			return false
		}
	}
	return true
}

// Watch blocks, watching for omarchy theme changes and calling emit with the
// new theme until ctx is done. It returns nil immediately if the omarchy
// directory is absent.
func Watch(ctx context.Context, emit func(models.OmarchyTheme)) error {
	dir, err := CurrentDir()
	if err != nil {
		return err
	}
	if st, err := os.Stat(dir); err != nil || !st.IsDir() {
		slog.Info("omarchy current dir not present; theme watcher inactive")
		return nil
	}
	return WatchDir(ctx, dir, emit)
}

// WatchDir watches dir (non-recursively) for any event, debounces, re-reads
// the theme, and emits only when it actually changed: the watcher fires for
// unrelated churn in the dir (e.g. background rotation touching symlinks).
// fsnotify does not report access events, so re-reading the theme files from
// inside the handler cannot re-trigger it (the loop the Rust build had to
// filter out).
func WatchDir(ctx context.Context, dir string, emit func(models.OmarchyTheme)) error {
	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		return fmt.Errorf("creating fs watcher: %w", err)
	}
	defer watcher.Close()
	if err := watcher.Add(dir); err != nil {
		return fmt.Errorf("watching %s: %w", dir, err)
	}
	slog.Info("watching omarchy theme dir", "dir", dir)
	read := func() models.OmarchyTheme {
		theme, err := ReadThemeFrom(dir)
		if err != nil {
			return KanagawaFallback()
		}
		return theme
	}
	last := read()
	for {
		select {
		case <-ctx.Done():
			return nil
		case err, ok := <-watcher.Errors:
			if !ok {
				return errors.New("fs watcher closed")
			}
			slog.Warn("omarchy watcher error", "error", err)
		case _, ok := <-watcher.Events:
			if !ok {
				return errors.New("fs watcher closed")
			}
			// Debounce: drain events that arrive within the settle window.
			timer := time.NewTimer(Debounce)
		drain:
			for {
				select {
				case <-watcher.Events:
					if !timer.Stop() {
						<-timer.C
					}
					timer.Reset(Debounce)
				case <-timer.C:
					break drain
				case <-ctx.Done():
					timer.Stop()
					return nil
				}
			}
			theme := read()
			if Equal(theme, last) {
				continue
			}
			last = theme
			emit(theme)
			slog.Info("emitted omarchy theme change", "theme", theme.Name)
		}
	}
}
